general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+               }
+            }
+         }
+      }
+      else
+      {
+         // check the nodes
+         Tfloat dd;
+         for (auto it=itb; it!=ite; it++)
+         {
+            dd=Dist(it->pt, bc.sp);
+            if (dd<bc.dist) // update bc
+            {
+               bc.pos=it-data.begin();
+               bc.dist=dd;
+               bc.level=level;
+            }
+         }
+      }
+   }
+
+   /// Returns distances and indices of the n closest points to a point pt.
+   void NNS(PointND& pt,const int& npoints,
+            typename std::vector<NodeND>::iterator itb,
+            typename std::vector<NodeND>::iterator ite,
+            size_t level,
+            std::vector< std::tuple<Tfloat,Tindex> > & res)
+   {
+      std::uint8_t dim=(std::uint8_t) (level%ndim);
+      size_t siz=ite-itb;
+      typename std::vector<NodeND>::iterator mtb=itb+siz/2;
+      if (siz>2)
+      {
+         // median is at itb+siz/2
+         level=level+1;
+         Tfloat R=std::get<0>(res[npoints-1]);
+         // check central one
+         Tfloat dd=Dist(mtb->pt, pt);
+         if (dd<R)
+         {
+            res[npoints-1]=std::make_tuple(dd,mtb->ind);
+            std::nth_element(res.begin(), res.end()-1, res.end());
+            R=std::get<0>(res[npoints-1]);
+         }
+         if ((pt.xx[dim]-R)>mtb->pt.xx[dim]) // look to the right only
+         {
+            NNS(pt, npoints, itb+siz/2+1, ite, level, res);
+         }
+         else if ((pt.xx[dim]+R)<mtb->pt.xx[dim]) // look to the left only
+         {
+            NNS(pt, npoints, itb, itb+siz/2, level,   res);
+         }
+         else  // check all
+         {
+            NNS(pt,npoints, itb+siz/2+1, ite, level, res); // right
+            NNS(pt,npoints, itb, itb+siz/2, level,   res); // left
+         }
+      }
+      else
+      {
+         Tfloat dd;
+         for (auto it=itb; it!=ite; it++)
+         {
+            dd=Dist(it->pt, pt);
+            if (dd< std::get<0>(res[npoints-1])) // update the list
+            {
+               res[npoints-1]=std::make_tuple(dd,it->ind);
+               std::nth_element(res.begin(), res.end()-1, res.end());
+            }
+         }
+      }
+   }
+
+   /// Finds the set of indices of points within a distance R of a point pt.
+   void FindNeighborPoints(PointND& pt, Tfloat R,
+                           typename std::vector<NodeND>::iterator itb,
+                           typename std::vector<NodeND>::iterator ite,
+                           size_t level,
+                           std::vector<Tindex> & res)
+   {
+      std::uint8_t dim=(std::uint8_t) (level%ndim);
+      size_t siz=ite-itb;
+      typename std::vector<NodeND>::iterator mtb=itb+siz/2;
+      if (siz>2)
+      {
+         // median is at itb+siz/2
+         level=level+1;
+         if ((pt.xx[dim]-R)>mtb->pt.xx[dim]) // look to the right only
+         {
+            RNS(pt, R, itb+siz/2+1, ite, level, res);
```

Comment:
Many apologies for the dumb question, but where is this method defined?

Reply:
Nice one! It is a recursive call to the current method. Corrected.